### PR TITLE
Support `:breaka[dd]`, `:breakd[el]`, and `:breakl[ist]`

### LINKED ISSUE #7186: Support `:breaka[dd]`, `:breakd[el]`, and `:breakl[ist]`
These allow you to interact with breakpoints

### PR DESCRIPTION
**Pull request by elazarcoh**
**What this PR does / why we need it**:
It adds support for interacting with breakpoints.

**Which issue(s) this PR fixes**
fixes #7186

**Special notes for your reviewer**:

### CODE REVIEW

File: src/cmd_line/commands/breakpoints.ts [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+    });
+    return vscode.window.showQuickPick(lines).then((selected) => {
+      if (selected) {
+        const [idWithHashtag] = selected.split('\t');
+        const id = parseInt(idWithHashtag.replace('#', ''), 10) - 1;
+        const breakpoint = breakpoints[id] as AnyBreakpoint;
+        if (breakpoint && breakpoint.location !== undefined) {
+          return vscode.window.showTextDocument(breakpoint.location.uri).then(
+            () => {
+              vimState.cursorStartPosition = breakpoint.location.range.start;
```
Comment (elazarcoh) [inline src/cmd_line/commands/breakpoints.ts:200]: @J-Fields couldn't get it to work while in the same file... Jumping to another file seems to be working.

File: src/cmd_line/commands/breakpoints.ts [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+}
+
+/**
+ * Add Breakpoint Command
+ */
+type AddBreakpointExpr = { expr: string };
+type AddBreakpointFunction = { function: string };
+type AddBreakpointFile = { line: number; file: string };
+type AddBreakpointHere = 'here';
+type AddBreakpoint =
```
Comment (J-Fields) [inline src/cmd_line/commands/breakpoints.ts:22]: I'd prefer a discriminated union here (with a `type` string enum). I think
```typescript
if (this.addBreakpoint.type === 'file') {...}
```
is cleaner than custom type guards in cases like this

File: src/cmd_line/commands/breakpoints.ts [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+
+/**
+ * Delete Breakpoint Command
+ */
+type DelBreakpointById = { id: number };
+type DelAllBreakpoints = 'all';
+type DelBreakpointFunction = { function: string };
+type DelBreakpointFile = { line: number; file: string };
+type DelBreakpointHere = 'here';
+type DelBreakpoint =
```
Comment (J-Fields) [inline src/cmd_line/commands/breakpoints.ts:89]: Discriminated union, as above

File: src/cmd_line/commands/breakpoints.ts [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+}
+
+/**
+ * List Breakpoints Command
+ */
+class ListBreakpointsCommand extends ExCommand {
+  public override isRepeatableWithDot: boolean = false;
+
+  async execute(vimState: VimState): Promise<void> {
+    type AnyBreakpoint = vscode.Breakpoint & vscode.SourceBreakpoint & vscode.FunctionBreakpoint;
```
Comment (J-Fields) [inline src/cmd_line/commands/breakpoints.ts:175]: Isn't this type equivalent to `vscode.Breakpoint`?

File: src/cmd_line/commands/breakpoints.ts [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+      } = b as AnyBreakpoint;
+      let entry = '';
+      entry += `#${i + 1}\t`;
+      entry += enabled ? '$(circle-filled)\t' : '$(circle-outline)\t';
+      entry += condition ? '$(debug-breakpoint-conditional)\t' : '\t';
+      entry += location ? `${path.basename(location.uri.fsPath)}:${location.range.start.line + 1}` : '';
+      entry += functionName ? `$(debug-breakpoint-function)${functionName}` : '';
+      return entry;
+    });
+    return vscode.window.showQuickPick(lines).then((selected) => {
```
Comment (J-Fields) [inline src/cmd_line/commands/breakpoints.ts:192]: Is there a reason we don't want to use `await` here?

File: src/cmd_line/commands/breakpoints.ts [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+        const id = parseInt(idWithHashtag.replace('#', ''), 10) - 1;
+        const breakpoint = breakpoints[id] as AnyBreakpoint;
+        if (breakpoint && breakpoint.location !== undefined) {
+          return vscode.window.showTextDocument(breakpoint.location.uri).then(
+            () => {
+              vimState.cursorStartPosition = breakpoint.location.range.start;
+            },
+          );
+        } else {
+          return;
```
Comment (J-Fields) [inline src/cmd_line/commands/breakpoints.ts:204]: I think both of these `else` blocks are unnecessary?

File: src/cmd_line/commands/breakpoints.ts (hunk 1 above)
Comment (J-Fields) [reply src/cmd_line/commands/breakpoints.ts:200]: Hmm... try setting `vimState.cursorStopPosition`?

**Review by J-Fields — CHANGES_REQUESTED** [review]
Thank you! Left a handful of mostly minor comments - I'll need to do another read and some manual testing, but this is looking good.

File: src/cmd_line/commands/breakpoints.ts (hunk 5 above)
Comment (elazarcoh) [reply src/cmd_line/commands/breakpoints.ts:192]: I've returned the promise, so it's essentially the same. But I'll do it, as it seems you prefer it this way. 👍

File: src/cmd_line/commands/breakpoints.ts (hunk 1 above)
Comment (elazarcoh) [reply src/cmd_line/commands/breakpoints.ts:200]: Works!

**Comment by elazarcoh** [comment]
Hi,
I resolved all of the issues, I think.

As a general workflow, who should click "Resolve conversation"?

File: src/cmd_line/commands/breakpoints.ts (hunk 4 above)
Comment (elazarcoh) [reply src/cmd_line/commands/breakpoints.ts]: not exactly. but I made it clearer now

**Comment by J-Fields** [comment]
> As a general workflow, who should click "Resolve conversation"?

You can resolve them most of the time - if there's something you're still unsure about (or think I might be unsure about, like if you disagree with a change I suggested), leave it unresolved

**Pull request merged**